test.py

```diff
@@ -563,16 +563,65 @@ def run_test(framework, cluster, repo_root):
         logger.info(msg, runtests_script, framework.name)
         raise CommandFailure(cmd_args)
 
+def setup_junit_xml():
+    """Try to download and import junit_xml"""
+    workdir = get_work_dir()
+    lib_dir = os.path.join(workdir, 'lib', 'python')
+    pip_cmd = ['pip', 'install', 'junit-xml', '-t', lib_dir]
+    subprocess.check_call(pip_cmd)
+    sys.path.append(lib_dir)
+    import junit_xml
+
+def emit_junit_xml(cluster_launch_attempts, frameworks):
+    """Write out all the test actions failures to a junit file for jenkins or
+    similar"""
+    import junit_xml
+    launch_fake_testcases = []
+    for launch_attempt in cluster_launch_attempts:
+        attempt_duration = launch_attempt.end_time - launch_attempt.start_time
+        fake_test = junit_xml.TestCase(launch_attempt.name,
+                                       elapsed_sec=attempt_duration)
+        if launch_attempt.launch_succeeded:
+            fake_test.stdout = "Launch worked"
+        else:
+            fake_test.add_failure_info("Launch failed")
+        launch_fake_testcases.append(fake_test)
+
+    launch_suite = junit_xml.TestSuite("Cluster launches",
+            launch_fake_testcases)
+
+    fake_suites = []
+    fake_suites.append(launch_suite)
+    for framework in frameworks:
+        framework_testcases = []
+        for action_name, action in framework.actions.items():
+            action_duration = action['finish'] - action['start']
+            fake_test = junit_xml.TestCase(action_name,
+                                           elapsed_sec=action_duration,
+                                           stdout = action['stdout'],
+                                           stderr = action['stderr'])
+            if not action['ok']:
+                message = action['error_message']
+                if not message:
+                    message = "%s failed" % action_name
+                fake_test.add_failure_info(message, action['error_output'])
+            framework_testcases.append(fake_test)
+        framework_suite = junit_xml.TestSuite("%s actions" % framework.name,
+                framework_testcases)
+        fake_suites.append(framework_suite)
+
+    with open("junit_testpy.xml", "w") as f:
+        junit_xml.TestSuite.to_file(f, fake_suites)
+
+
 def report_failed_actions():
-    "Do useful things with the recorded successful and failed actions"
+    """Do useful things with the recorded successful and failed actions"""
```

Comment:
"Triple quotes are used even though the string fits on one line. This makes it easy to later expand it."

Reasonable, I suppose.

I use mega as my qualifier for: "Don't change this, but I'd like to know why"